### DISCUSSION TITLE
How to make the track smoother and let the virtual car move along it?

### DISCUSSION THREAD

**Discussion by Lilly-yang** [prl-mushr/mushr#16] [Q&A] [answered]
Hi,

I am a beginner in Mushr. I am using the Global Planner to make the virtual car move along a preset track on an 8-shaped map. But I had some questions:
**1. How to add passing points？
2. How to make the track smoother?
3. How to let the virtual car move along a track？Is there a tutorial?**

Here is the 8-shaped map:
![track4](https://user-images.githubusercontent.com/39558023/117008352-e9d04f80-acea-11eb-859c-dfe4f1cc1852.png)

Since the OMPL algorithm is to find the shortest distance, and I don't know how to insert the passing points, I can only get the track by dividing it into two segments.
![track8_result_turning_radius_5_kernal_size_1_interpolation_density_80_validity_resolution_0 0005](https://user-images.githubusercontent.com/39558023/117007085-80037600-ace9-11eb-829e-36915d34c423.png)

![track8_result_2_turning_radius_5_kernal_size_1_interpolation_density_40_validity_resolution_0 0005](https://user-images.githubusercontent.com/39558023/117007414-ed170b80-ace9-11eb-8889-9eff6a642df4.png)

As you can see from the above figure, the tracks obtained are not smooth. The arrow often shifts (like drunk driving).

**Here are some key parameters:**

- Parameters for the service:
-- turning_radius = 5

- Parameters for the planner:
-- kernal_size = 1
-- interpolation_density = 80 or 40 (it depends on the length of the trajectory)
-- validity_redolution = 0.0005

A further question is, how to make a virtual car move automatically along the track I got? The result returned by Global Planner is a pair of coordinate and angle(see below), but what I understand is that steering the car requires a pair of speed and angle(2.0, 0.09), right? If so, how to convert these two pairs? Are there any related examples or tutorials?
![image](https://user-images.githubusercontent.com/39558023/117033088-586cd780-ad02-11eb-887b-5a3e6c126c32.png)

I really need your help. Thanks!

**Comment by schmittlema** [comment, marked as answer]
Hi Lilly!
You caught us at a bit of a bad time because right now we are in the process of switching global planners and writing a nice tracking controller to follow planned paths (questions 2/3) . Here was the original tutorial for this planner: https://github.com/prl-mushr/mushr-website/blob/master/content/tutorials/global_planner.md. I'm not sure if that will be super helpful for what you are doing though :/ .

 Smoothing can be done with shortcutting via the ompl [PathSimplifier](https://ompl.kavrakilab.org/classompl_1_1geometric_1_1PathSimplifier.html) either with the shortcutPath() or simplify() functions. 
 
 Could you elaborate on what you mean by "passing points"?

**Reply by Lilly-yang** [reply]
Hi Matt,
Thank you for your quick reply. :)

I have read the Global Planner tutorial, but it only provides a path planning tutorial, without further tutorial about how to control the car. So it is still not possible to directly use Global Planner to automatically control the movement of the car, right? I found a way to control the car in Intro to ROS tutorial. It needs (speed, angle), but what I got from Global Planner is (coordinates, angle). I am not sure if they can be used together and how complicated to use them.

I will try the smooth function you suggest. Thanks!

The ‘passing point’ is the point between the starting point and the goal point. Since my map is 8-type, when the starting point and goal point in the same 'o' of the 8-type map, the OMPL algorithm will return the shortest path and ignore the other half 'o'. I want to insert a passing point in another 'o' to allow the algorithm to return a complete 8-type path.

My purpose is to make the car move automatically with this 8-type map. So it is not essentially the shortest path planning problem. If Global Planner is unable to do so at present, I would be happy to know if you have other suggestions.

**Reply by schmittlema** [reply]
There is a way to follow a path with the mushr car but it currently requires writing a tracking cost function for [mushr_rhc](https://github.com/prl-mushr/mushr_rhc) which we are in the process of writing now. A simple alternative (probably easier) is to write a PID or pure pursuit path tracking controller. Here are some great [slides](https://courses.cs.washington.edu/courses/cse490r/19sp/site/resources/lec14_pid_pure_pursuit.pdf) on the topic from a course taught about mushr. Essentially you want to write a node that takes in a path and sends control signals (speed, steering angle) to `mux/ackermann_cmd_mux/input/navigation`.  Here is some [ starter code](https://github.com/prl-mushr/mushr_control) from that course that could be helpful, but it is not official mushr code. It has both algorithms amongst other implemented, but you may need to tweak the ROS api to fit your application.

The starter code from the course I think can take in a preset path (it has 5 presets currently). So you could create a path manually and use that for all runs. This could get around the passing point problem. Is there a particular reason you need the planner or would a hand-made path work just as well for your application?

**Reply by Lilly-yang** [reply]
The _PID and Pure Pursuit Control_ method sounds good. I am gonna try it. Thanks! :)

**But I don't have permission to access the [starter code](https://github.com/prl-mushr/mushr_control). It returned ‘404’. Could you please share it with me again?**

My ideal goal is as shown in the [video](https://www.youtube.com/watch?v=N_ntS3ywYuM&t=53s). But considering that my map is not big and my priority now is to get the car to run as quickly as possible for follow-up work, in fact, any method is ok for me.

**Reply by schmittlema** [reply]
Oops it was a private repo. Could you try now?

**Reply by Lilly-yang** [reply]
Yes, I can access it now. Thank you!!!

**Comment by precioso** [comment]
Hi Matt,

If I may jump into the discussion. What we try to do with Lilly is to automatize the real Mushr car to acquire data on the real track that we have virtualized:
![20200724_194608](https://user-images.githubusercontent.com/19729578/117481461-53f32980-af63-11eb-9791-7c5ac073f98c.jpg)
Our plan was to program virtually any path on the virtual track, then transfer the path into the real Mushr car to train deep networks from the data we will have obtained.

From what you describe and what we understand, it may be simpler to manipulate the car with the remote control, and forget for the moment the virtual simulation.

Thank you very much for your help.

Best,
fred

**Reply by schmittlema** [reply]
Cool setup! yeah sounds good. Appologies for the autonomy stack not being as effective as you want. We are in the process of revamping it to work better overall and should have something smooth. I should note, it looks like the car in your scene is not a mushr car, so the dynamics may be slightly different than our car if you use simulation. Please let me know if you have any further questions!

**Reply by precioso** [reply]
Yes this was just for the photo (and the memory of the first projects I was supervising as professor https://www.youtube.com/watch?v=-Nv1VqNL87s&ab_channel=RaspberryPi-Cars ) but we have built two mushr cars, and plan to have 3 or 4 more and to expand the track on demand.
We will look forward all the updates of mushr platform.
Thank you very much as always.
